### PR TITLE
explicit dependencies support added to package-build-package as an optional arg

### PR DESCRIPTION
**Pull request by lxsameer**
This way third party applications like **Cask** can provide a list of dependencies as mentioned in official **Emacs** docs to be included in the `-pkg.el` file in the tar packages or somewhere in the single files.

I need it to make some changes to **Cask** too.

### CODE REVIEW

File: package-build.el [hunk 1]
```diff
@@ -775,13 +775,15 @@ Optionally PRETTY-PRINT the data."
...
 
-(defun pb/merge-package-info (pkg-info name version)
+(defun pb/merge-package-info (pkg-info name version dependencies)
   "Return a version of PKG-INFO updated with NAME, VERSION and info from CONFIG.
 If PKG-INFO is nil, an empty one is created."
   (let* ((merged (or (copy-sequence pkg-info)
                      (vector name nil "No description available." version))))
     (aset merged 0 name)
     (aset merged 3 version)
+    (when (not (aref merged 1))
```
Comment (purcell) [inline package-build.el]: `unless` would be better here. :-)

**Comment by purcell** [comment]
This patch doesn't look unreasonable, but I'm wondering why nothing like this was necessary previously? `Cask` has been around for a long time, and was always able to generate packages without issues as far as I can tell.

File: package-build.el (hunk 1 above)
Comment (lxsameer) [inline package-build.el]: I can change it if you like

**Comment by lxsameer** [comment]
@purcell yeah you're right, but I want to make it better. currenly it create multi-file package without explicit dependencies even when this concept is part of the Emacs standard.  

Also I'm adding a `push` functionality to `cask` which requires clear dependencies in pkg file.

**Comment by milkypostman** [comment]
Is this part of the `package-build.el` spec somehow? Don't we already put dependencies in when we build a package? I mean, we at least have the dependencies in the `archive-contents` file?

Guess what I'm asking is: *why do we have to pass in the dependencies here when we already gather them from the package?*

**Comment by lxsameer** [comment]
Why not ? First of all it's good to have the option to override default functionality intentionally. Second you collect dependencies by looking at a exists pkg file. What if it doesn't exists. For example if you use Cask and don't create the pkg file. Cask already knows about dependencies.

File: package-build.el [hunk 2]
```diff
@@ -775,13 +775,15 @@ Optionally PRETTY-PRINT the data."
...
+      (aset merged 1 dependencies))
```
Comment (milkypostman) [inline package-build.el]: I'm not sure that it makes sense to *only* replace the dependencies if they don't exist. In fact if this is the case I don't think we should merge this.

For the behavior to make sense we should either make it so the dependencies are *union'd* or the dependencies are *replaced*.

As a first step I'm happy if you just replace the deps if they are specified.

```
(when dependencies
    (aset merged 1 dependencies))
```

and in the future we can discuss a merge.

Comment (lxsameer) [inline package-build.el]: done in new commit

**Comment by milkypostman** [comment]
@purcell I'm OK with this. I think the other possibility would be to build in some Cask support to our script to read deps from there and do the merge accordingly. but in some ways I prefer not having dependencies on other packages.

File: package-build.el [hunk 3]
```diff
@@ -1123,12 +1125,12 @@ Returns the archive entry for the package."
...
      ((= 1 (length files))
-      (pb/build-single-file-package package-name version (caar files) source-dir target-dir))
+      (pb/build-single-file-package package-name version (caar files) source-dir target-dir dependencies))
      ((< 1 (length  files))
-      (pb/build-multi-file-package package-name version files source-dir target-dir))
+      (pb/build-multi-file-package package-name version files source-dir target-dir dependencies))
      (t (error "Unable to find files matching recipe patterns")))))
 
-(defun pb/build-single-file-package (package-name version file source-dir target-dir)
+(defun pb/build-single-file-package (package-name version file source-dir target-dir dependencies)
```
Comment (purcell) [inline package-build.el]: I'm a little bit bothered here that the archive entry will contain the dependencies, but the modified `.el` package file will not: this means it can't be correctly installed locally with `package-install-file`. The file's `Package-Requires` header should probably get overwritten in this case, much as the `-pkg.el` file is rewritten in multi-file packages.

Comment (lxsameer) [inline package-build.el]: So you think we should remove dependencies for single packages right ?

Comment (purcell) [inline package-build.el]: Not sure yet whether we should replace or merge, but it should be consistent with whatever we decide for multi-file packages.

Comment (lxsameer) [inline package-build.el]: Personally I think we have to create a separate function which returns the dependencies in the `package-build` fashion. Then replace dependencies if the `dependencies` argument passed to `package-build-package`. 

This way user can manually fetch calculated dependencies and decide what to do with it. Merge or replace, then pass the resulted dependencies to `package-build-package`.

This way we won't need to write code for tools like **Cask** ( @milkypostman solution ), and we will have much clear API.

Comment (purcell) [inline package-build.el]: The problem with replacing is that if `dependencies` is nil, does that mean "leave the existing dependencies" or "there are no dependencies"?

Comment (lxsameer) [inline package-build.el]: I'm going with `leave the existing dependencies` because since the argument is optional changes of it being `nil` is high. Also as a developer if I want to override it with empty list of dependencies I simply can provide an empty list or vector ( this solution needs small changes ).

Comment (milkypostman) [inline package-build.el]: @purcell to your problem with an empty list being `nil`, we can just have `t` represent an empty replacement. However, in general this is more and more starting to feel like a hack that should be done a better way.

@lxsameer does Cask change any files in a package? I mean, it's my understanding that Cask writes version info into the file before it asks `package-build` to run? why doesn't it rewrite the dependencies?

Comment (lxsameer) [inline package-build.el]: @milkypostman by default it does not write dependencies anywhere. So In order to fix that I need this functionality.

Also I think merging this pull request would not hurt any one, and we can implement a better solution in the future.

Comment (purcell) [inline package-build.el]: But Cask *does* already write dependencies into the `-pkg.el` file it generates, right?

**Comment by purcell** [comment]
P.S. Is there a corresponding Cask issue for the improvements you're planning? I'm still struggling to understand exactly what helpful sequence of steps these changes would enable. I can't really comment on whether the changes are fit for purpose without understanding that purpose.

**Comment by lxsameer** [comment]
@purcell No there is not issue yet. I'm waiting for this one.

Basically **Cask** knows about dependencies via `Cask` file and when you create a package it does not store the dependencies in the pkg file. You have to build the pkg file before and put the dependencies inside it. 

With this patch we can pass dependencies directly at the packaging time in `cask-package` function and avoid all the steps before.

**Comment by milkypostman** [comment]
OK, here is my point of view: Cask is an alternative to the builtin mechanisms of `package.el`. That is, it replaces many of the keywords (package-version, requires, etc.) that already exist in somewhat separate and arbitrary locations (e.g., single file packages are in the comments, multi file have the `-pkg.el` file, etc.).

Thus, I think that we have two options as a team (`package-build.el` and Cask):

1. `package-build.el` adds support for reading the `Cask` file in packages; as support for *yet another* way of defining package versions and dependencies.

2. Cask does its job. If Cask's job is to manage dependencies and versions, then it should do that in the existing system that the Emacs community has defined.  That is, Cask should be generating and replacing dependencies and versioning information in the package that it is being used for.

I argue for 2 simply because it is actually the path of least resistance in general. Why? Because `package-build` already supports the "Emacs" way of doing things. Cask wants to add a layer of abstraction, which is fine with me, but then it should find a way to integrate that with the existing system. At some layer, unless `package.el` and the Emacs community decides that Cask format is the way to go, Cask has to switch from its way of doing things to the Emacs way of doing things. 

What you are asking us to do in this pull request feels like adding functionality that Cask should be doing, but in a way that is *general* and doesn't make sense in all contexts. In other words, this change is dependent on Cask's functionality, and not on the functionality we support: Emacs.


@lxsameer to your point about the `-pkg.el` file. It would be easy for Cask to generate this file before passing the package over to *package-build*. It would also be easy for Cask to inject the proper dependencies in the `.el` files. And my argument is that it **should**. Because what happens if the `.el` file lists *some* deps but then Cask adds more and the users installs the package. Then there are two sets of dependencies; the ones defined in the `.el` file and the ones defined in the Cask file which will not be accessible from the installed package.

**Comment by lxsameer** [comment]
I'm not the `cask` author. And I don't want to defend it or anything. As a developer I want to use the tool that makes my life easier. 

But about you philosophy about the `Emacs way` with all due respect I think your idea of Emacs way is not totally right. Containing solutions for third party developers to override what your software do is not a bad thing. Also there are many tools in free software world that do the same thing but for different tastes. 
I obviously can do this using **Cask** without changing package-build. But I have to do this in the hard way or even I can role my fork of `package-build` and `cask` and do what ever I want. But in this case it wouldn't be a community.
Also It's not an Iron rule which Emacs way of doing thing it totally true and people should love it. I think in this case a tool like **Cask** which use **Ruby Bundler** as a role model done a hell of a job and made the packaging and developing new Emacs lisp libraries much easier and fun. 

Anyway I think this patch would make `package-build` even better. If you're not agree with me you can simply reject this. It's up to you guys.

**Comment by milkypostman** [comment]
@lxsameer I just want to find a solution that makes sense.

**Comment by lxsameer** [comment]
@milkypostman dude it's easy . First we have to forget about the **Cask**. Then if user provides a list of dependencies for `package-build-package` the it should override the exists dependencies ( those which we calculated ) and if he passes a nil value we have to behave as before.

**Comment by purcell** [comment]
The job of `package-build-package` is "take this set of packaging-ready files and bundle them into an installable package with this version". Injecting additional dependencies changes this straightforward task significantly -- look at all the discussion about merging dependency info etc. That *might* be okay if it were clearly necessary, but it's currently rather speculative.

So my suggestion would be that we decline this PR for now, and you can propose the `cask` changes you'd like to see. In the context of those enhancements, if you and @rejeep find that changes to `package-build` are necessary, we will all have the context to understand how best to accomplish that.

We already put in a lot of effort to help make `package-build` usable by Cask, so we're certainly open to making changes driven by concrete need.

**Comment by lxsameer** [comment]
Ok then. 
Enhancing cask without this patch is usless.

So I go with forking both package build and cask.

**Comment by milkypostman** [comment]
@lxsameer i don't think your reaction makes sense in light of the conversation here. what @purcell is proposing is that you make a pull request to cask which assumes the change here so that we can see them both together and have a more enlightening conversation.

**Comment by lxsameer** [comment]
Also you guys almost do the same thing with name and version. So i think you should rethink that too.

**Comment by lxsameer** [comment]
@milkypostman dude don't get me wrong i don't have any hard feeling toward you guys. Your software your rules. I just want you guys to see my point.

**Comment by purcell** [comment]
> what @purcell is proposing is that you make a pull request to cask which assumes the change here so that we can see them both together and have a more enlightening conversation.

Yes, exactly. We're perfectly happy to make changes if we can be helped to understand why they are necessary and desirable. I don't think we quite got to that point in this PR, but since your main goal is to enhance Cask, let's start there, and feel free to pull us into any conversations about that.

**Comment by lxsameer** [comment]
As I mentioned you guys did the same thing in `package-build-package` for **package-name** and **version** which with you point of view is totally unnecessary and pointless. You can extract those info  easily from `pkg` file too. So why bother to pass them as arguments to function. You only needs the files list and the pkg file. 

So maybe you need to rethink that too.

**Comment by purcell** [comment]
`package-name` is not unnecessary because when there are multiple files in a package, there needs to be a way to indicate which the "primary" file is. We don't require there to be an existing `-pkg.el` for multi-file packages. But I do understand the point you're making.

**Comment by lxsameer** [comment]
@purcell  Correct me If I'm wrong. `package-build-package` will look for a `pkg` file in source three and extract the package data such as name, version, short description from it it it was present. If not it will get the data from the main el file ( which as you mentioned you need the package name for this purpose ).  let's call this step 1.
https://github.com/milkypostman/melpa/blob/master/package-build.el#L1176
 
Then override the name and version in pkg-info ( the data extracted from pkg or el file )  with those passed as arguments to `package-build-package` ( in merge-package-info ). step 2.

**My point**: You think this patch is not valid because I'm trying to override an element (dependencies) in package data with the one I'm passing to `package-build-package` . You think it does not follows the Emacs way of things. 

But as you can see in step 2 your are doing it with **name** and **version**. 
If you want to follow your philosophy ( the reason you argue with this patch ), then you shouldn't pass the name and version to `package-build-package` too. and pass the path to the `pkg` file or the main `el` file instead and extract the data from them. This way you don't have to override data too.

So long story short: you are arguing about a solution which you already implemented. If you think this patch in not valid then you have to change your code base too, If you think this is valid then why arguing ?

At last let me tell you this. I'm discussing this issue with only because I think `package-build` can benifit from this. And I'm not getting defensive or something.

**Comment by milkypostman** [comment]
Interesting observation about `name` and `version`.

So I am not sure about `name` but I know we do `version` because that is the design of melpa. We wanted packages to be versioned based on the VCS version and date was always the best way. Thus we definitely *want* to overwrite the version. 

Name needs to be specified because of what @purcell said. We need to know what files to look for. The reason *dependancies* is drastically different is because we want package owners to maintain this information.

**Comment by lxsameer** [comment]
@milkypostman dude I didn't mean that `name` and `version` are unnecessary in general. I said you can extract these from `pkg` or `el` file.

**Comment by purcell** [comment]
> I said you can extract these from pkg or el file.

Yes, but we effectively use the `name` part to identify the `-pkg.el` or main `.el` file from which to read metadata.

**Comment by lxsameer** [comment]
That's the point you can pass the manifest ( pkg or el file) path to the function instead of passing name version. right

**Comment by milkypostman** [comment]
Who is going to provide the file name? Yet another function. We're just
cutting out the middle man :)

**Comment by lxsameer** [comment]
@milkypostman When I want to use `package-build-package` to create a package I just need a `pkg` or `el` file. Am I right buddy ?
